sstables/storage.cc

```diff
@@ -114,6 +117,43 @@ future<> filesystem_storage::rename_new_file(const sstable& sst, sstring from_na
     });
 }
 
+future<> filesystem_storage::sync_cached_directory(std::string_view name) const {
+    std::string key(name);
```

Comment:
Can't the caller of this function keep, alongside `name`, the file object?

Reply:
That's what I'd like to know myself!

- The directory called `_dir.native()` is sync'd in the following `filesystem_storage` methods: `open()`, `create_links_common()`, `seal()`
- method `create_links_common()` also sync's a directory whose name it takes as a parameter (`dst_dir`), the owner of which appears difficult to find
- method `wipe()` sync's the _parent_ directory of whatever `make_toc_temporary()` returns
- method `atomic_delete_complete()` sync's the directory that's part of the passed-in `filesystem_atomic_delete_ctx`; the directory is likely determined when the context is originally constructed.
- method `move()` flushes the old directory (`_dir.native()`) and the new directory `new_dir` that's passed in as a parameter

Most of these methods -- that is, the direct callers of `flush` and `sync_directory` -- don't seem to "own" the directory that needs to be flushed. I guess we could introduce a new `filesystem_storage` field for keeping `_dir.native()` open, just for the sake of sync'ing, but I have no idea about the rest. "Opportunistic" caching via a hash table looked attractive because of this.

Reply:
The systematic way to approach it is to wrap both the directory name and its file object in something, and pass it instead of the directory name. Then you have both available.

We still have to determine its lifetime, but that's easy, it's tied to the table object.

Reply:
`sstable::snapshot()` is a public method. It takes `const sstring& dir` and forwards the call to `_storage->snapshot()` (asking for the directory to be interpreted as an absolute one).

`s3_storage::snapshot()` throws an exception (`Snapshotting S3 objects not implemented`).

`filesystem_storage::snapshot()` creates the snapshot directory, and calls `create_links_common()` on it. `create_links_common()` syncs the snapshot directory three times.

Indirectly, this exposes `sync_directory()` (`seastar/src/util/file.cc`) to any consumer of `sstable`, using a raw absolute directory name interface. (One further caller is `table::take_snapshot()` in `replica/table.cc`, which simiarly takes a raw directory name as a paramater.)

Are we supposed to change all these interfaces to `(sstring, file)` from just `sstring`?

Reply:
The challenge posted by snapshot() is that it's called for each sstable individually, while the directory is common to all of them.

Possible solutions:
1. Call snapshot with a container of sstables, rather than once per sstable. This requires a bit of surgery. The call chain is table::take_snapshot -> parallel loop -> sstable::snapshot -> storage::snapshot, it would have to change to table::take_snapshot -> storage::snapshot -> parallel_loop. Then storage::snapshot can open the directory once and sync it once.
2. Create a "token" for the directory and pass it around. table::take_snapshot() calls storage::make_snapshot_directory() once, storing the result, calls the existing callchain with the extra snapshot_directory parameter (replacing the string `dir`), this gets passed down to filesystem_storage::snapshot(),  which uses it to create the links. In the end, table::take_snapshot() calls snapshot_directory::commit() which syncs and close()s the file object it encapsulates.

IMO the second is easier.

Reply:
@avikivity OK, you are effectively confirming the global interface change.

(@xemul please help me with the following questions! Thanks!)

Because I have zero idea of the _semantics_ of `filesystem_storage`, I can only work this in a bottom-up fashion. That is, I need to locate each instance of `sync_directory`, and each instance of _temporary_ `open_checked_directory` / `open_directory`, and replace those with `flush` and `close` method calls on such `file` (directory) objects that _already exist_ (i.e., that were opened _permanently_, earlier). For each affected `filesystem_storage` method, update the parameter list such that the method take such a `file` object, rather than a directory name. Then, allow the compiler to report the damage, and propagate the interface change upwards, from callee towards caller.

For starting small, I've now exercised the above method on `filesystem_storage::atomic_delete_complete`. Revelent commits from the past, for this feature, are:

1. ed043e576222 ("sstable_directory: Split delete_with_pending_deletion_log()", 2023-12-01)
2. 4ecf4c4a6a7d ("sstables/storage: Add prepare/complete skaffold for atomic deletion", 2023-12-05)
3. 604279f064b8 ("sstables/storage: Reimplement atomic deletion in sstables_manager", 2023-12-05)
4. b9abd504be72 ("sstables/storage: Drop atomic deleter", 2023-12-05)

The first commit split off the preparation of the deletion log to a separate step, and introduced a context structure between (a) log prep, and (b) unlink sstables + log drop. The directory underlying the `filesystem_storage` object, at the time of preparation, was captured inside this context structure. **I do not understand why contextualizing the directory was necessary** (and the commit message doesn't say).

Commits 2 through 4 shifted the execution of atomic deletion from storage to manager. Regardless, the same context-passing remained in place, from (a) log prep, and (b) unlink sstables + log drop. That's what we have today in `master` (as of commit f31d5e3204f8).

In order to eliminate the `sync_directory` call from `filesystem_storage::atomic_delete_complete` (that is, to eliminate the ad-hoc opening of the directory, just for flushing the directory), I've now changed the context structure to carry an open `file` object, rather than a directory pathname, with the following patch:

~~~
diff --git a/sstables/sstable_directory.hh b/sstables/sstable_directory.hh
index c12b8501be71..1aa5e271563f 100644
--- a/sstables/sstable_directory.hh
+++ b/sstables/sstable_directory.hh
@@ -288,7 +288,7 @@ class sstable_directory {
     // Creates the deletion log for atomic deletion of sstables (helper for the
     // above function that's also used by tests)
     // Returns a pair of "logilfe name" and "directory with sstables"
-    static future<std::pair<sstring, sstring>> create_pending_deletion_log(const std::vector<shared_sstable>& ssts);
+    static future<std::pair<sstring, file>> create_pending_deletion_log(const std::vector<shared_sstable>& ssts);
 
     static bool compare_sstable_storage_prefix(const sstring& a, const sstring& b) noexcept;
 };
diff --git a/sstables/sstable_directory.cc b/sstables/sstable_directory.cc
index 6c945683e9cd..a6f3d5b01dd8 100644
--- a/sstables/sstable_directory.cc
+++ b/sstables/sstable_directory.cc
@@ -558,7 +558,7 @@ bool sstable_directory::compare_sstable_storage_prefix(const sstring& prefix_a,
     return size_a == size_b && sstring::traits_type::compare(prefix_a.begin(), prefix_b.begin(), size_a) == 0;
 }
 
-future<std::pair<sstring, sstring>> sstable_directory::create_pending_deletion_log(const std::vector<shared_sstable>& ssts) {
+future<std::pair<sstring, file>> sstable_directory::create_pending_deletion_log(const std::vector<shared_sstable>& ssts) {
     return seastar::async([&ssts] {
         shared_sstable first = nullptr;
         min_max_tracker<generation_type> gen_tracker;
@@ -583,6 +583,7 @@ future<std::pair<sstring, sstring>> sstable_directory::create_pending_deletion_l
         sstring pending_delete_dir = first->_storage->prefix() + "/" + sstables::pending_delete_dir;
         sstring pending_delete_log = format("{}/sstables-{}-{}.log", pending_delete_dir, gen_tracker.min(), gen_tracker.max());
         sstring tmp_pending_delete_log = pending_delete_log + ".tmp";
+        auto prefix_dir_f = open_directory(first->_storage->prefix()).get();
         sstlog.trace("Writing {}", tmp_pending_delete_log);
         try {
             touch_directory(pending_delete_dir).get();
@@ -615,7 +616,7 @@ future<std::pair<sstring, sstring>> sstable_directory::create_pending_deletion_l
             sstlog.warn("Error while writing {}: {}. Ignoring.", pending_delete_log, std::current_exception());
         }
 
-        return std::make_pair<sstring, sstring>(std::move(pending_delete_log), first->_storage->prefix());
+        return std::make_pair<sstring, file>(std::move(pending_delete_log), std::move(prefix_dir_f));
     });
 }
 
diff --git a/sstables/storage.cc b/sstables/storage.cc
index e3a0b92ad1f9..0559c4c52831 100644
--- a/sstables/storage.cc
+++ b/sstables/storage.cc
@@ -473,8 +473,8 @@ future<> filesystem_storage::wipe(const sstable& sst, sync_dir sync) noexcept {
 class filesystem_atomic_delete_ctx : public atomic_delete_context_impl {
 public:
     sstring log;
-    sstring directory;
-    filesystem_atomic_delete_ctx(sstring l, sstring dir) noexcept : log(std::move(l)), directory(std::move(dir)) {}
+    file dir_f;
+    filesystem_atomic_delete_ctx(sstring l, file dir) noexcept : log(std::move(l)), dir_f(std::move(dir)) {}
 };
 
 future<atomic_delete_context> filesystem_storage::atomic_delete_prepare(const std::vector<shared_sstable>& ssts) const {
@@ -485,7 +485,8 @@ future<atomic_delete_context> filesystem_storage::atomic_delete_prepare(const st
 future<> filesystem_storage::atomic_delete_complete(atomic_delete_context ctx_) const {
     auto& ctx = static_cast<filesystem_atomic_delete_ctx&>(*ctx_);
 
-    co_await sync_directory(ctx.directory);
+    co_await ctx.dir_f.flush();
+    co_await ctx.dir_f.close();
 
     // Once all sstables are deleted, the log file can be removed.
     // Note: the log file will be removed also if unlink failed to remove
~~~

All (local) unit tests pass.

My questions:

1. Regardless of _representation_ of the base directory to sync (i.e., pathname versus open file) in the context structure, **why** does the base directory have to be part of the context at all? Why don't we just sync `_dir` upon completion, just like we used to, before commit 1 (ed043e576222, "sstable_directory: Split delete_with_pending_deletion_log()", 2023-12-01)?

    Do we expect the `_dir` member of `filesystem_storage` to possibly change between (a) log prep, and (b) unlink sstables + log drop? For example, due to an intervening call to `filesystem_storage::change_state --> filesystem_storage::move` (while `sstables_manager::delete_atomically` is processing the unlink operations in parallel)? For which reason the context needs to preserve the original base directory?

2. Does the transformation shown above **actually buy us anything**?

    Both before and after the patch, we open the base directory precisely once, just for the sake of flushing. The effect of the transformation is only that we keep the directory open for a longer time, before we flush it and close it. I don't see how this benefits us.

3. Now, if this particular patch is indeed useless, then that's a problem: if I need to eliminate `sync_directory` calls on a case-by-case basis from `filesystem_storage` (dependent on whether the given elimination will save us open/close calls), then we'll be left with a _mixture_ of one-off `sync_directory` calls, and open-coded sequences of `dir_f.flush()`. Is that an improvement for the codebase?

Reply:
General question - haven't we passed the line where it makes sense to design this first in a document, then go back to the implementation when the agreed upon design is reviewed and approved?

Reply:
We sync the directory multiple times during its lifetime: each time we link in an sstable, each time we delete (one or a set of) sstables. But we need to open and close it only once.

Before this blows up into a huge project, let's identify the simplest case where we can eliminate the open, and do it just there.

Reply:
I'll work on that, thanks.

For improving my understanding of the code in general though, I'd still like to get an answer to my [question 1](https://github.com/scylladb/scylladb/pull/19624#discussion_r1671988363) -- why does the base directory have to be part of the atomic deletion context? @xemul @tchaikov can you help with that please? Thanks.

Reply:
I don't know the answer, but if I have to guess, it's a generalization. sstables don't live in a single directory, there's upload/ and staging/ (and snapshots/*/).

Reply:
Correct. There is the base directory common to all sstables in a single table.

Please see https://github.com/scylladb/scylladb/pull/19555
In particular, the first patch: [`2df4d51` (#19555)](https://github.com/scylladb/scylladb/pull/19555/commits/2df4d51230803c3a9e620810a67fc2eb2b81786f)

Reply:
@bhalevy thank you.

Reply:
> let's identify the simplest case where we can eliminate the open, and do it just there

I'm attempting to do that in v4.